services/cargohold/test/integration/API.hs

```diff
@@ -22,324 +21,40 @@ module API (tests) where
-  -- Lookup and download via redirect.
-  r2 <-
-    downloadAsset uid loc () <!! do
-      const 302 === statusCode
-      const Nothing === responseBody
-  r3 <- flip get' id =<< parseUrlThrow (C8.unpack (getHeader' "Location" r2))
-  liftIO $ do
-    assertEqual "status" HTTP.status200 (responseStatus r3)
-    assertEqual "content-type mismatch" (Just applicationOctetStream) (getContentType r3)
-    assertEqual "user mismatch" uid (decodeHeaderOrFail "x-amz-meta-user" r3)
-    assertEqual "data mismatch" (Just "test") (responseBody r3)
-  where
-    exampleMultipart :: LByteString
-    exampleMultipart =
-      "--FrontierIyj6RcVrqMcxNtMEWPsNpuPm325QsvWQ\r\n\
-      \Content-Type: application/json;charset=utf-8\r\n\
-      \Content-length: 37\r\n\
-      \\r\n\
-      \{\"public\":true,\"retention\":\"eternal\"}\r\n\
-      \--FrontierIyj6RcVrqMcxNtMEWPsNpuPm325QsvWQ\r\n\
-      \Content-Type: application/octet-stream\r\n\
-      \Content-length: 4\r\n\
-      \Content-MD5: CY9rzUYh03PK3k6DJie09g==\r\n\
-      \\r\n\
-      \test\r\n\
-      \--FrontierIyj6RcVrqMcxNtMEWPsNpuPm325QsvWQ--\r\n\
-      \\r\n"
-
 --------------------------------------------------------------------------------
 -- Federation behaviour
 
-testRemoteDownloadWrongDomain :: TestM ()
-testRemoteDownloadWrongDomain = do
-  assetId <- liftIO $ Id <$> nextRandom
-  uid <- liftIO $ Id <$> nextRandom
-
-  let key = AssetKeyV3 assetId AssetPersistent
-      qkey = Qualified key (Domain "invalid.example.com")
-  downloadAsset uid qkey () !!! do
-    const 422 === statusCode
-
-testRemoteDownloadNoAsset :: TestM ()
-testRemoteDownloadNoAsset = do
-  assetId <- liftIO $ Id <$> nextRandom
-  uid <- liftIO $ Id <$> nextRandom
-  let key = AssetKeyV3 assetId AssetPersistent
-      qkey = Qualified key (Domain "faraway.example.com")
-      respond req
-        | frRPC req == "get-asset" =
-            pure ("application" // "json", Aeson.encode (GetAssetResponse False))
-        | otherwise =
-            throw
-              . MockErrorResponse HTTP.status404
-              . LText.decodeUtf8With Text.lenientDecode
-              . Aeson.encode
-              $ assetNotFound
-  (_, reqs) <- withMockFederator respond $ do
-    downloadAsset uid qkey () !!! do
-      const 404 === statusCode
-  localDomain <- viewFederationDomain
-  liftIO $
-    reqs
-      @?= [ FederatedRequest
-              { frOriginDomain = localDomain,
-                frTargetDomain = Domain "faraway.example.com",
-                frComponent = Cargohold,
-                frRPC = "get-asset",
-                frBody = Aeson.encode (GetAsset uid key Nothing)
-              }
-          ]
-
+-- This test doesn't fit in well with the `integration` style of tests. This is
+-- because setting up the asset to have the weird state where the asset key exists,
+-- but somehow the asset itself can't be downloaded is a bit beyond what the API
+-- integration setup allows. So this specific test can stay here for now.
```

Comment:
I've added a small mocking utility to `integration`: https://github.com/wireapp/wire-server/pull/3742/commits/75889a5c30bff06d3885705b8c497f3afe53c80f. Maybe we can extend it to support this use case. In another PR, though.

Reply:
Ah, that is very interesting then. I can look into that, but yes a new ticket once that is merged I think.